# Modeling

In this notebook, we train machine learning models to predict delivery time.

We use the preprocessed training data to train the models and the test data to evaluate their performance.


In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression

## Load Preprocessed Data

We load the training and test datasets created during preprocessing.


In [ ]:
X_train = pd.read_csv("../data/X_train_preprocessed.csv")
X_test = pd.read_csv("../data/X_test_preprocessed.csv")

y_train = pd.read_csv("../data/y_train.csv").squeeze() # squeeze() converts the one-column DataFrame into a Series
y_test = pd.read_csv("../data/y_test.csv").squeeze()

## Verify Loaded Data

We check the shapes of the loaded datasets before training the model.


In [3]:
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (33217, 28)
X_test: (8305, 28)
y_train: (33217,)
y_test: (8305,)


## Linear Regression

Linear Regression is a supervised learning algorithm used to predict a numerical value.

We use it as our first baseline model to predict `Time_taken(min)`.


In [4]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

## Train the Model

We train the Linear Regression model using the training features and their corresponding delivery times.


In [10]:
model.fit(
    X_train,
    y_train
)

print("Model trained successfully.")

Model trained successfully.


In [7]:
print(model)

LinearRegression()


## Linear Regression — Evaluation

After training the Linear Regression model, we evaluate its performance on the test set using several regression metrics:

* **MAE**: average prediction error in minutes.
* **MSE**: average squared prediction error.
* **RMSE**: prediction error in minutes, with more importance given to large errors.
* **R²**: proportion of the variation in delivery time explained by the model.
* **Adjusted R²**: R² adjusted according to the number of features used by the model.

These metrics allow us to evaluate the model from different perspectives and compare it with other regression models.


In [26]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Predictions
y_pred = model.predict(X_test)

# Metrics
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

# Adjusted R²
n = len(y_test)       # number of observations
p = X_test.shape[1]   # number of features

adjusted_r2 = 1 - (1 - r2) * (n - 1) / (n - p - 1)

# Display results
print("Linear Regression Results")
print("-------------------------")
print("MAE        :", round(mae, 2), "minutes")
print("MSE        :", round(mse, 2))
print("RMSE       :", round(rmse, 2), "minutes")
print("R²         :", round(r2, 4))
print("Adjusted R²:", round(adjusted_r2, 4))

Linear Regression Results
-------------------------
MAE        : 5.07 minutes
MSE        : 41.22
RMSE       : 6.42 minutes
R²         : 0.5362
Adjusted R²: 0.5346


## Decision Tree Regression

Decision Tree Regression predicts a numerical value by splitting the data into smaller groups using feature values.

We use it as a second model to compare its performance with Linear Regression.


In [18]:
from sklearn.tree import DecisionTreeRegressor

tree_model = DecisionTreeRegressor(
    random_state=42
)

## Train the Decision Tree

We train the Decision Tree using the training features and the corresponding delivery times.


In [19]:
tree_model.fit(
    X_train,
    y_train
)

print("Decision Tree trained successfully.")

Decision Tree trained successfully.


## Make Predictions

We use the trained Decision Tree to predict delivery times for the test data.


In [20]:
tree_pred = tree_model.predict(X_test)

print("First 10 predictions:")
print(tree_pred[:10])

First 10 predictions:
[25. 24. 17. 41. 24. 43. 19. 15. 23. 47.]


## Evaluate the Decision Tree

We evaluate the Decision Tree using MAE, MSE, RMSE, and R².

These metrics allow us to compare the Decision Tree with our Linear Regression baseline.


In [27]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Predictions
tree_pred = tree_model.predict(X_test)

# Metrics
tree_mae = mean_absolute_error(y_test, tree_pred)
tree_mse = mean_squared_error(y_test, tree_pred)
tree_rmse = np.sqrt(tree_mse)
tree_r2 = r2_score(y_test, tree_pred)

# Adjusted R²
n = len(y_test)
p = X_test.shape[1]

tree_adjusted_r2 = 1 - (1 - tree_r2) * (n - 1) / (n - p - 1)

# Display results
print("Decision Tree Results")
print("---------------------")
print("MAE        :", round(tree_mae, 2), "minutes")
print("MSE        :", round(tree_mse, 2))
print("RMSE       :", round(tree_rmse, 2), "minutes")
print("R²         :", round(tree_r2, 4))
print("Adjusted R²:", round(tree_adjusted_r2, 4))

Decision Tree Results
---------------------
MAE        : 4.83 minutes
MSE        : 41.28
RMSE       : 6.43 minutes
R²         : 0.5355
Adjusted R²: 0.534


## Compare the Baseline Models

The Decision Tree and Linear Regression models have similar performance on the test set.

The Decision Tree slightly reduces MAE, while Linear Regression has slightly better MSE, RMSE, and R².

These results show that neither model has a clear advantage at this stage.


## Random Forest Regression

Random Forest is an ensemble regression algorithm that combines multiple Decision Trees.

Using several trees can improve generalization compared with a single Decision Tree and can help reduce overfitting.


In [22]:
from sklearn.ensemble import RandomForestRegressor

rf_model = RandomForestRegressor(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

## Train the Random Forest

We train the Random Forest using the training data.

The test data remains unseen during training.


In [23]:
rf_model.fit(
    X_train,
    y_train
)

print("Random Forest trained successfully.")

Random Forest trained successfully.


## Make Predictions

We use the trained Random Forest to predict delivery times for the test data.


In [24]:
rf_pred = rf_model.predict(X_test)

print("First 10 predictions:")
print(rf_pred[:10])

First 10 predictions:
[23.59 24.18 16.93 40.09 23.65 39.65 15.12 21.74 25.56 47.79]


## Evaluate the Random Forest

We evaluate the Random Forest using MAE, MSE, RMSE, and R².

These metrics will later be compared with the other candidate models.


In [28]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Predictions
rf_pred = rf_model.predict(X_test)

# Metrics
rf_mae = mean_absolute_error(y_test, rf_pred)
rf_mse = mean_squared_error(y_test, rf_pred)
rf_rmse = np.sqrt(rf_mse)
rf_r2 = r2_score(y_test, rf_pred)

# Adjusted R²
n = len(y_test)
p = X_test.shape[1]

rf_adjusted_r2 = 1 - (1 - rf_r2) * (n - 1) / (n - p - 1)

# Display results
print("Random Forest Results")
print("---------------------")
print("MAE        :", round(rf_mae, 2), "minutes")
print("MSE        :", round(rf_mse, 2))
print("RMSE       :", round(rf_rmse, 2), "minutes")
print("R²         :", round(rf_r2, 4))
print("Adjusted R²:", round(rf_adjusted_r2, 4))

Random Forest Results
---------------------
MAE        : 3.67 minutes
MSE        : 22.25
RMSE       : 4.72 minutes
R²         : 0.7496
Adjusted R²: 0.7488


### Interpretation

The Random Forest model achieves a **MAE of 3.67 minutes**, which means that the model's predictions are wrong by approximately **3.67 minutes on average**.

The model also achieves an **RMSE of 4.72 minutes** and an **R² of 0.7496**, meaning that it explains approximately **74.96% of the variation in delivery time** on the test set.

Compared with Linear Regression and a single Decision Tree, Random Forest provides better test-set performance. It is therefore a strong candidate for the next step: **hyperparameter optimization**.
